File: 2013 University of Chicago Invitational Programming Contest/winterroads.py
import sys
from collections import deque, defaultdict

class DSU:
    def __init__(self, n):
        self.parent = list(range(n + 1))
    
    def find(self, i):
        if self.parent[i] == i:
            return i
        self.parent[i] = self.find(self.parent[i])
        return self.parent[i]
    
    def unite(self, i, j):
        root_i = self.find(i)
        root_j = self.find(j)
        if root_i != root_j:
            self.parent[root_i] = root_j
            return True
        return False

def main():
    data = sys.stdin.read().split()
    idx = 0
    
    n = int(data[idx]); idx += 1
    m = int(data[idx]); idx += 1
    
    adj = defaultdict(list)
    all_edges = [None] * (m + 1)
    capacities = [0] * (m + 1)
    
    initial_edges = []
    for i in range(1, m + 1):
        u = int(data[idx]); idx += 1
        v = int(data[idx]); idx += 1
        cap = int(data[idx]); idx += 1
        all_edges[i] = (u, v, i, cap)
        capacities[i] = cap
        initial_edges.append(all_edges[i])
    
    # Build initial MSF
    initial_edges.sort(key=lambda x: -x[3])
    dsu = DSU(n)
    
    for u, v, r_id, cap in initial_edges:
        if dsu.unite(u, v):
            adj[u].append((v, r_id))
            adj[v].append((u, r_id))
    
    e = int(data[idx]); idx += 1
    
    def check_path(a, b, w):
        visited = [False] * (n + 1)
        queue = deque([a])
        visited[a] = True
        
        while queue:
            u = queue.popleft()
            if u == b:
                return True
            for v, r_id in adj[u]:
                if not visited[v] and capacities[r_id] >= w:
                    visited[v] = True
                    queue.append(v)
        return False
    
    def remove_from_adj(u, v, r_id):
        adj[u] = [(neighbor, rid) for neighbor, rid in adj[u] if not (neighbor == v and rid == r_id)]
        adj[v] = [(neighbor, rid) for neighbor, rid in adj[v] if not (neighbor == u and rid == r_id)]
    
    def find_path_bottleneck(start, end):
        visited = [False] * (n + 1)
        parent = [0] * (n + 1)
        road_to_parent = [0] * (n + 1)
        queue = deque([start])
        visited[start] = True
        found = False
        
        while queue and not found:
            u = queue.popleft()
            for v, r_id in adj[u]:
                if not visited[v]:
                    visited[v] = True
                    parent[v] = u
                    road_to_parent[v] = r_id
                    queue.append(v)
                    if v == end:
                        found = True
                        break
        
        if not found:
            return (-1, -1)
        
        min_cap = float('inf')
        min_road_id = -1
        curr = end
        
        while curr != start:
            r_id = road_to_parent[curr]
            if capacities[r_id] < min_cap:
                min_cap = capacities[r_id]
                min_road_id = r_id
            curr = parent[curr]
        
        return (min_cap, min_road_id)
    
    def add_edge(u, v, r_id):
        adj[u].append((v, r_id))
        adj[v].append((u, r_id))
    
    for _ in range(e):
        type_char = data[idx]; idx += 1
        if type_char == 'S':
            a = int(data[idx]); idx += 1
            b = int(data[idx]); idx += 1
            w = int(data[idx]); idx += 1
            print(1 if check_path(a, b, w) else 0)
        else:
            r_id = int(data[idx]); idx += 1
            c_new = int(data[idx]); idx += 1
            c_old = capacities[r_id]
            capacities[r_id] = c_new
            u, v, _, _ = all_edges[r_id]
            
            was_in_msf = any(neighbor == v and rid == r_id for neighbor, rid in adj[u])
            
            if was_in_msf and c_new < c_old:
                remove_from_adj(u, v, r_id)
                
                # Find component containing u
                in_T_u = [False] * (n + 1)
                queue = deque([u])
                in_T_u[u] = True
                
                while queue:
                    curr = queue.popleft()
                    for neighbor, _ in adj[curr]:
                        if not in_T_u[neighbor]:
                            in_T_u[neighbor] = True
                            queue.append(neighbor)
                
                # Find best replacement edge
                max_cap = -1
                best_r_id = -1
                for j in range(1, m + 1):
                    uj, vj, _, cap_j = all_edges[j]
                    if in_T_u[uj] != in_T_u[vj]:
                        if capacities[j] > max_cap:
                            max_cap = capacities[j]
                            best_r_id = j
                
                if best_r_id != -1:
                    add_edge(all_edges[best_r_id][0], all_edges[best_r_id][1], best_r_id)
            
            elif not was_in_msf and c_new > c_old:
                min_cap, min_road_id = find_path_bottleneck(u, v)
                if min_cap != -1 and c_new > min_cap:
                    remove_from_adj(all_edges[min_road_id][0], all_edges[min_road_id][1], min_road_id)
                    add_edge(u, v, r_id)

if __name__ == "__main__":
    main()

File: 2013 University of Chicago Invitational Programming Contest/winterroads.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Edge {
    public int u, v, id;
    public long cap;
}

public class DSU {
    private int[] parent;
    
    public DSU(int n) {
        parent = new int[n + 1];
        for (int i = 0; i <= n; i++) {
            parent[i] = i;
        }
    }
    
    public int Find(int i) {
        if (parent[i] == i) return i;
        return parent[i] = Find(parent[i]);
    }
    
    public bool Unite(int i, int j) {
        int root_i = Find(i);
        int root_j = Find(j);
        if (root_i != root_j) {
            parent[root_i] = root_j;
            return true;
        }
        return false;
    }
}

public class Program {
    static List<(int, int)>[] adj;
    static Edge[] allEdges;
    static long[] capacities;
    static int n, m, e;
    
    static bool CheckPath(int a, int b, long w) {
        bool[] visited = new bool[n + 1];
        Queue<int> queue = new Queue<int>();
        queue.Enqueue(a);
        visited[a] = true;
        
        while (queue.Count > 0) {
            int u = queue.Dequeue();
            if (u == b) return true;
            
            foreach (var (v, rId) in adj[u]) {
                if (!visited[v] && capacities[rId] >= w) {
                    visited[v] = true;
                    queue.Enqueue(v);
                }
            }
        }
        return false;
    }
    
    static void RemoveFromAdj(int u, int v, int rId) {
        adj[u].RemoveAll(x => x.Item1 == v && x.Item2 == rId);
        adj[v].RemoveAll(x => x.Item1 == u && x.Item2 == rId);
    }
    
    static (long, int) FindPathBottleneck(int start, int end) {
        bool[] visited = new bool[n + 1];
        int[] parent = new int[n + 1];
        int[] roadToParent = new int[n + 1];
        Queue<int> queue = new Queue<int>();
        
        queue.Enqueue(start);
        visited[start] = true;
        bool found = false;
        
        while (queue.Count > 0 && !found) {
            int u = queue.Dequeue();
            
            foreach (var (v, rId) in adj[u]) {
                if (!visited[v]) {
                    visited[v] = true;
                    parent[v] = u;
                    roadToParent[v] = rId;
                    queue.Enqueue(v);
                    if (v == end) {
                        found = true;
                        break;
                    }
                }
            }
        }
        
        if (!found) return (-1, -1);
        
        long minCap = long.MaxValue;
        int minRoadId = -1;
        int curr = end;
        
        while (curr != start) {
            int rId = roadToParent[curr];
            if (capacities[rId] < minCap) {
                minCap = capacities[rId];
                minRoadId = rId;
            }
            curr = parent[curr];
        }
        
        return (minCap, minRoadId);
    }
    
    static void AddEdge(int u, int v, int rId) {
        adj[u].Add((v, rId));
        adj[v].Add((u, rId));
    }
    
    public static void Main() {
        string[] input = Console.ReadLine().Split();
        n = int.Parse(input[0]);
        m = int.Parse(input[1]);
        
        adj = new List<(int, int)>[n + 1];
        for (int i = 0; i <= n; i++) adj[i] = new List<(int, int)>();
        allEdges = new Edge[m + 1];
        capacities = new long[m + 1];
        
        List<Edge> initialEdges = new List<Edge>();
        for (int i = 1; i <= m; i++) {
            input = Console.ReadLine().Split();
            allEdges[i] = new Edge {
                id = i,
                u = int.Parse(input[0]),
                v = int.Parse(input[1]),
                cap = long.Parse(input[2])
            };
            capacities[i] = allEdges[i].cap;
            initialEdges.Add(allEdges[i]);
        }
        
        // Build initial MSF
        initialEdges = initialEdges.OrderByDescending(e => e.cap).ToList();
        DSU dsu = new DSU(n);
        
        foreach (var edge in initialEdges) {
            if (dsu.Unite(edge.u, edge.v)) {
                AddEdge(edge.u, edge.v, edge.id);
            }
        }
        
        e = int.Parse(Console.ReadLine());
        for (int i = 0; i < e; i++) {
            input = Console.ReadLine().Split();
            char type = input[0][0];
            
            if (type == 'S') {
                int a = int.Parse(input[1]);
                int b = int.Parse(input[2]);
                long w = long.Parse(input[3]);
                Console.WriteLine(CheckPath(a, b, w) ? "1" : "0");
            } else {
                int rId = int.Parse(input[1]);
                long cNew = long.Parse(input[2]);
                long cOld = capacities[rId];
                capacities[rId] = cNew;
                int u = allEdges[rId].u;
                int v = allEdges[rId].v;
                
                bool wasInMsf = adj[u].Any(x => x.Item1 == v && x.Item2 == rId);
                
                if (wasInMsf && cNew < cOld) {
                    RemoveFromAdj(u, v, rId);
                    
                    // Find component containing u
                    bool[] inTU = new bool[n + 1];
                    Queue<int> queue = new Queue<int>();
                    queue.Enqueue(u);
                    inTU[u] = true;
                    
                    while (queue.Count > 0) {
                        int curr = queue.Dequeue();
                        foreach (var (neighbor, _) in adj[curr]) {
                            if (!inTU[neighbor]) {
                                inTU[neighbor] = true;
                                queue.Enqueue(neighbor);
                            }
                        }
                    }
                    
                    // Find best replacement edge
                    long maxCap = -1;
                    int bestRId = -1;
                    for (int j = 1; j <= m; j++) {
                        if (inTU[allEdges[j].u] != inTU[allEdges[j].v]) {
                            if (capacities[j] > maxCap) {
                                maxCap = capacities[j];
                                bestRId = j;
                            }
                        }
                    }
                    
                    if (bestRId != -1) {
                        AddEdge(allEdges[bestRId].u, allEdges[bestRId].v, bestRId);
                    }
                } else if (!wasInMsf && cNew > cOld) {
                    var (minCap, minRoadId) = FindPathBottleneck(u, v);
                    if (minCap != -1 && cNew > minCap) {
                        RemoveFromAdj(allEdges[minRoadId].u, allEdges[minRoadId].v, minRoadId);
                        AddEdge(u, v, rId);
                    }
                }
            }
        }
    }
}